### ISSUE TITLE
How to reference a value when defining a new value?

### ISSUE THREAD

**Issue by pmorch** [k14s/ytt#309]
Newbie here.

# TL;DR

How do I create a new value in an `overlay.yaml` file achieving this:

```yaml
serviceA_URL: #@ data.values.base_URL + "/service/A"
```

so that it later can be used in the tempate as `data.values.serviceA_URL`?

# Details

Assume we have a number of REST endpoints. Normally all these REST endpoints are available on the same server in a default place, but sometimes, e.g. during development, some of the endpoints are available in a completely different place.

For 95% of cases I have `values.yaml`:

```yaml
#@data/values
---
base_URL: https://staging-server
```

and want output:

```yaml
endpoints:
  serviceA_URL: https://staging-server/service/A
  serviceB_URL: https://staging-server/service/B
  serviceC_URL: https://staging-server/service/C
```

But sometimes I want this `values.yaml` instead:

```yaml
#@data/values
---
base_URL: https://staging-server
serviceB_URL: http://dev-server/some/other/path/B
```

and then get this output:

```yaml
endpoints:
  serviceA_URL: https://staging-server/service/A
  serviceB_URL: http://dev-server/some/other/path/B
  serviceC_URL: https://staging-server/service/C
```

I'd prefer it if my final template doesn't know that the different `serviceX_URL` URLs often are are related. So I was thinking of having `values.yaml` as the input, and then having an `overlay.yaml` to define the values for  `serviceX_URL`  based on  the ["Multiple Data Values" playground example](https://carvel.dev/ytt/#example:example-multiple-data-values) and then have this as the final `config.yaml`:

```yaml
#@ load("@ytt:data", "data")
endpoints:
  serviceA_URL: #@ data.values.serviceA_URL
  serviceB_URL: #@ data.values.serviceB_URL
  serviceC_URL: #@ data.values.serviceC_URL
```

I was hoping then to have an `overlay.yaml` something like:
```yaml
#@ load("@ytt:overlay", "overlay")
#@data/values
---
#@ load("@ytt:data", "data")
#@overlay/match missing_ok=True
serviceA_URL: #@ data.values.base_URL + "/service/A"
#@overlay/match missing_ok=True
serviceB_URL: #@ data.values.base_URL + "/service/B"
#@overlay/match missing_ok=True
serviceC_URL: #@ data.values.base_URL + "/service/C"
```

But it seems I can't access `data.values` from `overlay.yaml`. I started out with the ["Multiple Data Values" playground example](https://carvel.dev/ytt/#example:example-multiple-data-values) and thought I could just add `#@ load("@ytt:data", "data")` to the mix, but it doesn't work.

The error I get is:

```bash
$ ytt -f values.yaml  -f overlay.yaml -f config.yaml 
ytt: Error: Overlaying data values (in following order: values.yaml, overlay.yaml):
  Templating file 'overlay.yaml': 
- NoneType has no .base_URL field or method
    in <toplevel>
      overlay.yaml:6 | serviceA_URL: #@ data.values.base_URL + "/service/A"
```

How can I define new values in `overlay.yaml` based on existing values?

And if that can be done, is there some way I can only set a value if it hasn't been set already in `values.yaml`? I'm thinking something like `#@overlay/only-if-missing=True` instead of `#@overlay/match missing_ok=True`? Or would `if` statements be the way to go?

I really want an overlay on top of `values.yaml` and not on top of the final output, because in reality, the  `serviceX_URL` values are used in many places, and I'd like my code to be as DRY as possible.

**Comment by cari-lynn** [comment]
Hey @pmorch, thanks for opening this question!

I see a need to input a base_URL that can be used for all endpoints, and be able to modify any individual endpoints that may need to change.

Currently, accessing `data.values` is not supported until after all data values files have been processed, so it's not possible to share `base_URL` across data values files. We put together an alternative approach that allows you to define a default `base` URL as well as individual service URLs and paths.

Data values file `_defaults.yaml` defines a `default` endpoint base URL that can be used by all services, as well as individual service base and paths for when those need to be changed.
```
#@data/values
---
endpoints:
  #! default base of all endpoints
  default:
    base: https://default-server

  serviceA:
    base: ~
    path: /service/v2/A

  serviceB:
    base: ~
    path: /service/B

  serviceC:
    base: ~
    path: /service/C
```

Define a function that takes data.values and the service name as inputs, applies the default 'base' URL if no base for that service is defined, and returns the concatenated base and path. This function makes it possible to have both base and path parameterized as inputs in the data values file, but present them as one string.

Note: this can be put in the template as well, putting the function in a `.star` file allows us to omit the `#@` that begins a line of starlark code in a yaml file.

`endpoints.star` file:
```
def service_url(dvs, name):
   base = dvs[name].base or dvs.default.base
   return base + dvs[name].path
end
```

Use this function in your template to `config.yaml`
Note: We are aliasing `d = data.values` for brevity.
```
#@ load("@ytt:data", "data"); d = data.values
#@ load("endpoints.star", "service_url")

---
endpoints:
  serviceA_URL: #@ service_url(d.endpoints, "serviceA")
  serviceB_URL: #@ service_url(d.endpoints, "serviceB")
  serviceC_URL: #@ service_url(d.endpoints, "serviceC")
  ```
 This results in output:
```
endpoints:
  serviceA_URL: https://default-server/service/v2/A
  serviceB_URL: https://default-server/service/B
  serviceC_URL: https://default-server/service/C
```

Lastly, if you want to modify the base URL or a single endpoint, you can overlay just the values you want to modify in `z-data-values.yml`. 
Note: The name of the file matters since `ytt` orders inputs alphabetically, it starts with a `z` because we want this overlay to be processed last.
```
#@data/values
---
endpoints:
  default:
    base: https://staging-server

  serviceA:
    base: http://new-server
    path: /as/new/path
```
[Here's](https://carvel.dev/ytt/#gist:https://gist.github.com/cari-lynn/5d6f4a4a8969199d479d38e051258b35) a link to the solution in the playground

I believe now that the solution is different than before some of your questions may have been addressed, please let me know if this solution doesn't solve your use case or answer all your questions, I'm happy to help.

**Comment by pmorch** [comment]
Hi Cari,

Thank you very much for your well thought-out answer. I know these take time to write, and I appreciate it! 

> Currently, accessing data.values is not supported until after all data values files have been processed, so it's not possible to share base_URL across data values files.

This was the original question, and "not possible" is a valid answer to it. :-(

For the specifics of my question, your workaround works great. I've come up with another that works for us too:

```yaml
#@ base_URL = "https://staging-server"
endpoints:
  serviceA_URL: #@ base_URL + "/serviceA_URL"
  serviceB_URL: #@ base_URL + "/serviceB_URL"
  serviceC_URL: #@ base_URL + "/serviceC_URL"
```

I like your solution better, actually.

But really, what I was looking for was a generic way of referencing a `data.values` key in an overlay.
